packages/statemanager/src/statefulVerkleStateManager.ts

```diff
@@ -0,0 +1,239 @@
+    this.originalStorageCache = new OriginalStorageCache(this.getStorage.bind(this))
+    this._caches = opts.caches
+    this.keccakFunction = opts.common?.customCrypto.keccak256 ?? keccak256
+    this.verkleCrypto = opts.verkleCrypto
+  }
+
+  getAccount = async (address: Address): Promise<Account | undefined> => {
+    const elem = this._caches?.account?.get(address)
+    if (elem !== undefined) {
+      return elem.accountRLP !== undefined ? createAccountFromRLP(elem.accountRLP) : undefined
+    }
+
+    const stem = getVerkleStem(this.verkleCrypto, address, 0)
+
+    // First retrieve the account "header" values from the trie
+    const accountValues = await this._trie.get(stem, [
+      VerkleLeafType.BasicData,
+      VerkleLeafType.CodeHash,
+    ])
+
+    let account
+    if (accountValues[0] !== undefined) {
+      console.log(accountValues)
+      const basicData = decodeVerkleLeafBasicData(accountValues[0]!)
+      account = createPartialAccount({
+        version: basicData.version,
+        balance: basicData.balance,
+        nonce: basicData.nonce,
+        codeHash: accountValues[1] instanceof Uint8Array ? accountValues[1] : null,
+        codeSize: basicData.codeSize,
+        storageRoot: null, // TODO: Add storage stuff
+      })
+    }
+    // check if the account didn't exist if any of the basic keys are undefined
+    else if (accountValues[0] === undefined || accountValues[1] === undefined) {
+      if (this.DEBUG) {
+        this._debug(`getAccount address=${address.toString()} from DB (non-existent)`)
+      }
+      this._caches?.account?.put(address, account)
+    }
+
+    if (this.DEBUG) {
+      this._debug(`getAccount address=${address.toString()} stem=${short(stem)}`)
+    }
+    return account
+  }
+
+  putAccount = async (address: Address, account?: Account): Promise<void> => {
+    if (this.DEBUG) {
+      this._debug(
+        `Save account address=${address} nonce=${account?.nonce} balance=${
+          account?.balance
+        } contract=${account && account.isContract() ? 'yes' : 'no'} empty=${
+          account && account.isEmpty() ? 'yes' : 'no'
+        }`,
+      )
+    }
+    if (this._caches?.account === undefined) {
+      if (account !== undefined) {
+        const stem = getVerkleStem(this.verkleCrypto, address, 0)
+        const basicDataBytes = encodeVerkleLeafBasicData({
+          version: account.version,
+          balance: account.balance,
+          nonce: account.nonce,
+          codeSize: account.codeSize,
+        })
+        console.log(basicDataBytes)
+        await this._trie.put(
+          stem,
+          [VerkleLeafType.BasicData, VerkleLeafType.CodeHash],
+          [basicDataBytes, account.codeHash],
+        )
+      } else {
+        // Delete account
+      }
+    } else {
+      if (account !== undefined) {
+        this._caches?.account?.put(address, account, true)
+      } else {
+        this._caches?.account?.del(address)
+      }
+    }
+  }
```

Comment:
Don't want to get yourself too much on the nerve with this refactoring/DRY stuff (but maybe a bit 😂), but maybe you can put such code parts like the `putAccount()` directly in a `src/helpers/verkle.ts` file or similar with an adjusted function signature `putAccount(self/this: SM, address: Address, account?: Account): Promise<void>` taking the code from the stateless verkle SM directly over so that these kind of code parts just never reach this file?

Then we directly have a unified version and do not need to post-DRY this and this should not be significantly ore work (just run the statless tests from `statelessVerkleStateManager.spec.ts` along)!

Reply:
I guess I'm not 100% sure what you mean since the implementations are not the same.  The `stateless` SM puts things in an in memory `map` via direct `this._state[verkleKey] = serializedBytes` and the `stateful` implementation uses `await this.trie.put(stem, suffixes, values)` to "put" data.  The thing I could do differently is to have `encodeVerkleLeafBasicData` accept an `account` rather than the various fields and that would de-dup a few lines of code.

Reply:
Ok, yeah, nevermind, you are right, haven't looked/though closely enough. I think I am generally still latently unsatisfied that we always copy over 2/3 of code which is redundant/the same.

Maybe we can also tackle this in a more fine-grained way. We could e.g. centralize the generic DEBUG strings and re-use, things like that (for sure not in this PR).

So don't let me distract you here! 🙂 I will nevertheless not give up on the topic, maybe open some more low-reaching small issues on this after this PR got merged.